### PR TITLE
Switch to merges in the upload scripts

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This follows on from the discussion in #94. It implements option 1:

> [and this is what I think you're suggesting] is that we change both the V1 and V2 upload artifacts operations to use a MERGE. That way we prevent the creation of new duplicates within V1 and within V2 at the point of upload [potentially causing errors if there are pre-existing duplicates however].

The upshot of this is that duplicates should be prevented on upload.

### CODE REVIEW

File: macros/upload_artifacts.sql [hunk 1]
```diff
@@ -21,19 +21,31 @@
...
+        merge into {{ src_dbt_artifacts }} as old_data using (
+            select
+            $1 as data,
+            $1:metadata:generated_at::timestamp_ntz as generated_at,
+            metadata$filename as path,
+            regexp_substr(metadata$filename, '([a-z_]+.json)') as artifact_type
+            from  @{{ src_dbt_artifacts }}
+        ) as new_data
+        -- Matching on the data payload itself. That's probably the most efficient.
+        on old_data.data = new_data.data
```
Comment (NiallRees) [inline macros/upload_artifacts.sql]:
```suggestion
        on old_data.generated_at = new_data.generated_at
```

Let's use `generated_at` to be consistent with the unique key specified in the dedupe script? Or is there a reason we can't?

File: macros/upload_artifacts_v2.sql [hunk 2]
```diff
@@ -42,13 +43,46 @@
             args:target::string as target,
             metadata,
             args
-        from raw_data;
+        from raw_data
+    ) as new_data
+    on old_data.artifact_run_id = new_data.artifact_run_id and old_data.artifact_generated_at = new_data.artifact_generated_at
```
Comment (NiallRees) [inline macros/upload_artifacts_v2.sql]:
```suggestion
    on old_data.artifact_run_id = new_data.artifact_run_id
```

I don't think the existing is bad - but again just for consistency with the dedupe script we should just be able to use artifact_run_id here, right?

File: macros/upload_artifacts_v2.sql [hunk 3]
```diff
@@ -63,13 +97,51 @@
 
         )
 
-        {{ flatten_results("raw_data") }};
+        {{ flatten_results("raw_data") }}
+
+    ) as new_data
+    on old_data.artifact_run_id = new_data.artifact_run_id and old_data.artifact_generated_at = new_data.artifact_generated_at
```
Comment (NiallRees) [inline macros/upload_artifacts_v2.sql]:
```suggestion
    on old_data.artifact_run_id = new_data.artifact_run_id and old_data.node_id = new_data.node_id
```

We have a composite key of artifact_run_id and node_id here as it's flattened right?

File: macros/upload_artifacts_v2.sql [hunk 4]
```diff
@@ -84,7 +156,56 @@
 
         )
 
-        {{ flatten_manifest("raw_data") }};
+        {{ flatten_manifest("raw_data") }}
+
+    ) as new_data
+    on old_data.artifact_run_id = new_data.artifact_run_id and old_data.artifact_generated_at = new_data.artifact_generated_at
```
Comment (NiallRees) [inline macros/upload_artifacts_v2.sql]:
```suggestion
    on old_data.artifact_run_id = new_data.artifact_run_id and old_data.node_id = new_data.node_id
```

File: macros/upload_artifacts_v2.sql (hunk 2 above)
Comment (alanmcruickshank) [reply macros/upload_artifacts_v2.sql]: In line with #103 and #94 - we should use `command_invocation_id` (+/- `node_id`) for these merges.

File: macros/upload_artifacts_v2.sql (hunk 3 above)
Comment (alanmcruickshank) [reply macros/upload_artifacts_v2.sql]: In line with #103 and #94 - we should use `command_invocation_id` (+/- `node_id`) for these merges.

File: macros/upload_artifacts_v2.sql (hunk 4 above)
Comment (alanmcruickshank) [reply macros/upload_artifacts_v2.sql]: In line with #103 and #94 - we should use `command_invocation_id` (+/- `node_id`) for these merges.

File: macros/upload_artifacts.sql [hunk 5]
```diff
@@ -21,19 +21,31 @@
...
+        -- Merge to avoid duplicates
+        merge into {{ src_dbt_artifacts }} as old_data using (
+            select
+            $1 as data,
+            $1:metadata:generated_at::timestamp_ntz as generated_at,
+            metadata$filename as path,
+            regexp_substr(metadata$filename, '([a-z_]+.json)') as artifact_type
+            from  @{{ src_dbt_artifacts }}
+        ) as new_data
+        -- Matching on the data payload itself. That's probably the most efficient.
```
Comment (NiallRees) [inline macros/upload_artifacts.sql]:
```suggestion
```

**Review by NiallRees — APPROVED** [review]

**Comment by danflippo** [comment]
I have kind of already written this logic. In my own situation, I am putting much of the project's logic into a stored procedure that is called by a very stripped down upload macro because one of my customers will have many projects across a large organization uploading logs to one place and we don't want to track down every dbt user when something changes. In my procedure, I use the following three sections near the beginning of my three insert statements to avoid duplicates. It isn't a merge but effectively avoids the duplicates on initial insert which is the main goal. Since I do this on upload, my later queries are all views because I don't have to do an expensive deduplication.

I am pasting some of my code from my stored procedure below because right now the rest of my logic differs from the project's latest DDL changes. If you agree with the approach, we should be able to drop this into the three insert statements relatively easily by replacing the table references with jinja and having later SQL reference deduplicated_data. I will try to get around to a pull request if I can but feel free to incorporate this in the meantime.

### Logic from DBT_RUN_RESULTS Insert:
```sql
with raw_data as (
    select
        $1:metadata as metadata,
        $1 as data
    from @` + STAGE_LOCATION + `/run_results.json.gz
    ),
deduplicated_data as (
    select
        metadata,
        data,
        metadata:invocation_id::string as command_invocation_id,
        -- NOTE: DBT_CLOUD_RUN_ID is case sensitive here 
        metadata:env:DBT_CLOUD_RUN_ID::int as dbt_cloud_run_id,
        coalesce(dbt_cloud_run_id::string, command_invocation_id::string) as artifact_run_id,
        metadata:generated_at::timestamp_tz as generated_at
    from raw_data
    where not exists (
        select null
        from DBT_RUN_RESULTS_NODES existing_rows
        where existing_rows.artifact_generated_at = metadata:generated_at::timestamp_tz
            and (existing_rows.dbt_cloud_run_id = metadata:env:DBT_CLOUD_RUN_ID::int
                or existing_rows.command_invocation_id = metadata:invocation_id::string)
    )
)
```
### Logic from DBT_RUN_RESULTS_NODES Insert:
```sql
with raw_data as (
    select
        $1:metadata as metadata,
        $1 as data
    from @` + STAGE_LOCATION + `/run_results.json.gz
    ),
deduplicated_data as (
    select
        metadata,
        data,
        metadata:invocation_id::string as command_invocation_id,
        -- NOTE: DBT_CLOUD_RUN_ID is case sensitive here 
        metadata:env:DBT_CLOUD_RUN_ID::int as dbt_cloud_run_id,
        coalesce(dbt_cloud_run_id::string, command_invocation_id::string) as artifact_run_id,
        metadata:generated_at::timestamp_tz as generated_at
    from raw_data
    where not exists (
        select null
        from DBT_RUN_RESULTS_NODES existing_rows
        where existing_rows.artifact_generated_at = metadata:generated_at::timestamp_tz
            and (existing_rows.dbt_cloud_run_id = metadata:env:DBT_CLOUD_RUN_ID::int
                or existing_rows.command_invocation_id = metadata:invocation_id::string)
    )
)
```

### Logic from DBT_MANIFEST_NODES Insert:
```sql
with raw_data as (
    select
        $1:metadata as metadata,
        $1 as data
    from @` + STAGE_LOCATION + `/manifest.json.gz
), 
deduplicated_data as (
    select
        metadata,
        metadata:invocation_id::string as command_invocation_id,
        -- NOTE: DBT_CLOUD_RUN_ID is case sensitive here 
        metadata:env:DBT_CLOUD_RUN_ID::int as dbt_cloud_run_id,
        coalesce(dbt_cloud_run_id::string, command_invocation_id::string) as artifact_run_id,
        metadata:generated_at::timestamp_tz as generated_at,
        data
    from raw_data
    where not exists (
        select null
        from DBT_MANIFEST_NODES existing_rows
        where existing_rows.artifact_generated_at = metadata:generated_at::timestamp_tz
            and (existing_rows.dbt_cloud_run_id = metadata:env:DBT_CLOUD_RUN_ID::int
                or existing_rows.command_invocation_id = metadata:invocation_id::string)
    ) 
)
```

**Pull request merged**